metals/src/main/scala/scala/meta/internal/metals/BuildTargets.scala

```diff
@@ -255,6 +261,32 @@ final class BuildTargets(
     }
   }
 
+  def addWorkspaceBuildSources(result: SourcesResult): Unit = {
+    val generatedDirs = result
+      .getItems()
+      .asScala
+      .flatMap { items =>
+        items
+          .getSources()
+          .asScala
+          .collect {
+            case item
+                if (item.getKind() == SourceItemKind.DIRECTORY && item
+                  .getGenerated()) =>
+              AbsolutePath(Paths.get(new URI(item.getUri())))
+          }
+          .toList
+      }
+      .toList
+    buildTargetGeneratedDirs = generatedDirs
+  }
```

Comment:
There is `addSourceItem` and `sourceItemToBuildTarget` in `BuildTargets` that already stores the info about sources.
It would be better to extend it's data  and adapt other methods that uses this field:
```scala

case class SourceItemInfo(buildTargets: List[BuildTarget], isGenerated: Boolean)

private val sourceItemsToBuildTarget = // need to rename
  TrieMap.empty[AbsolutePath, SourceItemInfo]
```

Reply:
- :+1: to register `SourceItem` to `BuildTargets` via `addSourceItem` instead of `addWorkspaceBuildSources`
  - I didn't realize there's a method to add `sourceItems`, nice catch!
- Yeah, it would be nice to store `isGenerated` in `SourceItemInfo` :+1:
  - because as @Arthurm1 mentioned, it's nice to see the `isGenerated` information in `SourceDirectories` section of `BuildTargetInfo` (of FileDecodeProvider).
  - https://github.com/scalameta/metals/blob/6c7d33f05eae64dd36f8c161dc2115f3a4c23c6f/metals/src/main/scala/scala/meta/internal/metals/BuildTargetInfo.scala#L221-L227
- However, I don't think we should **lookup** `isGenerated` information in that way because:
  - We should be able to prefix match against generated source dirs (and `TrieMap[AbsolutePath, SourceItemInfo]` can't) because, at the point of `import build` time, we might not have generated **files** yet.
    - Therefore, we register the list of directories where generated files will be stored, and exclude files that belong to those directories. (by prefix match)
  - Also, I would like to minimize the size of `buildTargetGeneratedDirs` as the prefix match algorithm is now `O(N)`, and it would be better to reduce the N.
    - Though we can optimize the prefix match by Trie or aho-corasick or something, I guess it's still worth reducing the size of the data structure.